### PR TITLE
Scrape our own prometheus metrics and inject them into builtin tables

### PR DESCRIPTION
**Pull request by antifuchs**
This PR builds out [this design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/20210311_prometheus_source.md) for #1696.

It introduces the following new relations as log tables:
- `mz_catalog.mz_metrics_meta` - meta-information about a metric (type and help text keyed by metric name).
- `mz_catalog.mz_metrics` - gauge and counter values from prometheus
- `mz_catalog.mz_metric_histograms` - readings from prometheus histograms, in a merge-able format (two arrays containing upper bounds and counts).

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/6271)

### CODE REVIEW

**Review by cirego — COMMENTED** [review]
This is very exciting and I'm looking forward to using this!

Random thought: are there any metrics that are sufficiently stable that we could write a testdrive test to verify the metrics over time and/or perhaps even the time-based expiry?

**Comment by antifuchs** [comment]
alright, I think this is ready to review for real. I hope to benchmark it still, but the general shape of the PR is basically there.

File: src/coord/src/coord/prometheus.rs
Comment (quodlibetor) [inline src/coord/src/coord/prometheus.rs:9]: could you add a module-level comment explaining that this mod injects metrics from the prometheus registry into our metrics tables?

File: src/coord/src/coord/prometheus.rs
Comment (quodlibetor) [inline src/coord/src/coord/prometheus.rs:115]: nit: normally we write what we do expect, not what caused the expectation to fail: `.expect("system clock after 1970")`, and the same in the expect below.

File: src/dataflow/src/logging/materialized.rs
Comment (quodlibetor) [inline src/dataflow/src/logging/materialized.rs:466]: This seems likely to create blips where metadata disappears and reappears? Do we have any other alternatives for this?

File: src/coord/src/coord/prometheus.rs
Comment (antifuchs) [reply src/coord/src/coord/prometheus.rs:115]: Haha, I think I cargo-culted that from elsewhere, but happy to fix here (:

File: src/dataflow/src/logging/materialized.rs
Comment (antifuchs) [reply src/dataflow/src/logging/materialized.rs:466]: Agreed, this is not as useful for metadata as for the regular metrics. I'll see if we can revoke the entries manually once they need to go away.

**Comment by antifuchs** [comment]
I also notice that the data in the histograms table is fairly useless for further processing: You can UNNEST each array separately, but you can't do them together because they're disjoint types, and so there's no way to associate each histogram bound with its respective count.

Maybe this should record a single row per histogram bucket instead.

**Comment by antifuchs** [comment]
Updated this to have a better histogram representation (one row per bucket bound, instead of keeping the bucket bounds & counts on separate arrays, which make usage difficult).

Now, you can do:

```sql
create view sql_command_duration_histo as 
   SELECT metric, labels, time, bound, count 
     FROM mz_metric_histograms outer_mmh 
   WHERE time = (SELECT max(time) FROM mz_metric_histograms max_mmh WHERE max_mmh.metric = outer_mmh.metric and max_mmh.labels=outer_mmh.labels);
```

to get the latest histogram, and query with:

```
materialize=> select bound, repeat('■'::text, (count::integer / (select sum(count) from sql_command_duration_histo) * 90)::integer) from sql_command_duration_histo order by bound;
  bound   |      repeat
----------+------------------
 0.000016 |
 0.000032 |
 0.000064 |
 0.000128 |
 0.000256 |
 0.000512 | ■■
    0.001 | ■■
    0.002 | ■■■■
    0.004 | ■■■■■
    0.008 | ■■■■■
    0.016 | ■■■■■
    0.032 | ■■■■■
    0.064 | ■■■■■
    0.128 | ■■■■■■■■■■
    0.256 | ■■■■■■■■■■■■■■■■
    0.512 | ■■■■■■■■■■■■■■■■
        1 | ■■■■■■■■■■■■■■■■
(17 rows)
```

**Review by cirego — APPROVED** [review]
Awesome! Looks great to me but others may have more comments.

**Review by ruchirK — COMMENTED** [review]
I'm really sorry for the super late review asking for large changes Andreas :(

I don't really follow the approach of having these three relations exist as `BuiltinLogs`. It seems like we have a bunch of data that we extract within the `Scraper` thread, and then we broadcast that data to worker 0, and then worker 0 does the work to transform it into a set of `(Row, time, diff)` updates and then insert them into a system log. It seems like if these were `BuiltinTables` instead, then the scraper itself could do the manipulation to convert the data and metadata into `(Row, time, diff)` formats and then send those into the BuiltinTable directly -- using something similar to `Coordinator::update_catalog_view` (but adapted a little bit to support the retractions and stuff)

yeah basically I think having these as `BuiltinLog`s adds a bunch of complexity because the logic to manage everything gets split out amongst a bunch of different places. I think of `BuiltinLog`s as making sense when its dataflow workers writing down something about their own habits and activities, and tables as making sense when its the coordinator dictating what to write down.

I don't think the changes to do this should be terrible. Hopefully its mostly code movement into the `Scraper`? Also, let me know if you already considered this and it won't work for some reason that I'm blindly missing. it is late and I've done worse reading comprehension fails :D

sorry again for the late review.

File: src/coord/src/coord.rs
Comment (antifuchs) [reply src/coord/src/coord.rs:3520]: I tried making it one, but failed for some reason that I don't remember anymore. I can try again, but modeling it after the way the timestamper does it (mpsc channels & std threads) worked on the first try.

**Comment by antifuchs** [comment]
Point well taken on making it a BuildinTable vs. a BuiltinLog - I mostly chose the latter because I understood how to make those into things that are available on the catalog already, vs. having to research where they should go. Definitely happy to put them somewhere they fit better.

**Comment by antifuchs** [comment]
OK! I've updated this to latest `main` & reworked things according to @ruchirK's suggestion - I think this looks somewhat nicer now & is more tractable (fewer cross-crate/module dependencies!)

**Comment by antifuchs** [comment]
One thing that I'm not super happy with is that this does inserts at several timestamps (one per batch of metric kinds plus one per metadata batch), but we might not really need transaction-like semantics here. If you think this is necessary, I can adjust the code probably - right now it does make a decent-ish tradeoff between memory consumption per batch and transactionality.

File: src/coord/src/coord.rs
Comment (ruchirK) [inline src/coord/src/coord.rs:2906]: I haven't thought about this carefully but what if we just ... didn't join on drop? it seems like the `Scraper` thread needs this `tx` so that the `Coordinator` can tell it to shut down but the `Coordinator` needs to tell it to shut down because the coordinator blocks until the Scraper shuts down, but if we revoke that, then we don't need to do any of the extra stuff as a consequence. I think this is also a place where making the Scraper an async task would help (but that can wait till after this pr)

File: src/dataflow/src/logging/materialized.rs
Comment (ruchirK) [inline src/dataflow/src/logging/materialized.rs:30]: hmm why did this change?

**Review by ruchirK — APPROVED** [review]
This looks good to me! I haven't done a fully deep read of coord/prometheus yet but everything seems reasonable. I have one minor comment about making the interaction between the coordinator and the scraper simpler which i think can totally wait till after we land this.

as a sanity check -- have we run this against perf tests to make sure the default settings don't kill performance on load tests / the single threaded laptop experience?

File: src/dataflow/src/logging/materialized.rs
Comment (antifuchs) [reply src/dataflow/src/logging/materialized.rs:30]: It's kinda a remnant of the metrics-as-logging-tables change: I needed to pass things that don't impl these traits (f64, ugh), and realized that these traits are entirely unused & don't need to be there, so removed them. I can rebase that away if necessary, but the next person who needs something like an f64 on a message field might be happy it's gone (:

File: src/coord/src/coord.rs
Comment (antifuchs) [reply src/coord/src/coord.rs:2906]: Agreed that this is a bit more involved than it needs for the happy path - I tried to stay as true as I could to the timestamper design there; the thing that might make it worth it for now is tests, I think: We spawn a bunch of coordinators and having each of those leave threads sitting around that can panic might not be great (though I guess we can just drop the tx channel instead of sending a message? That might make this whole mechanism less involved).

For async tasks, I think it's gonna be difficult. The coord doesn't run an executor, so we'd have to figure out how to fit all that together. LMK what you think - happy either way.

Comment (ruchirK) [reply src/coord/src/coord.rs:2906]: lets merge this in rn and then async-ify / figure out how to simplify the coordinator as a followup (but can you make a task so we don't forget?)

File: src/dataflow/src/logging/materialized.rs
Comment (ruchirK) [reply src/dataflow/src/logging/materialized.rs:30]: thats fine then

**Comment by antifuchs** [comment]
(rebased & added release notes and a little cli docs snippet. This should merge cleanly!) (:

**Pull request merged**
